aten/src/ATen/native/SegmentReduce.cpp

```diff
@@ -103,21 +103,114 @@ void _segment_reduce_cpu_kernel1(
+  for (int64_t d = 0; d < axis; d++)
+      outer_offset *= output.size(d);
+  for (int64_t d = axis + 1; d < output.dim(); d++)
+      inner_offset *= output.size(d);
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      kBFloat16, kHalf, data.scalar_type(), "_segment_reduce_cpu", [&]() {
+        auto* output_data = output.data_ptr<scalar_t>();
+        const auto* values_data = data.data_ptr<scalar_t>();
+        for (const auto outer_idx : c10::irange(outer_offset)) {
+          int64_t lengths_cum_sum = 0;
+          for (const auto dim_idx : c10::irange(segment_count)) {
+            // FIXME: pass lengths
+            int64_t segment_length = lengths_data[outer_idx * lengths_stride_axis * segment_count + dim_idx];
+            for (const auto inner_idx : c10::irange(inner_offset)) {
+              // ===== step1: initialize starting value
+              // FIXME: could we just fill out with the inits beforehand?
+              scalar_t initial_value;
+              if (initial.has_value()) {
+                initial_value = initial.value().to<scalar_t>();
+              } else if (reduction == SegmentReductionType::MAX) {
+                initial_value = -std::numeric_limits<scalar_t>::infinity();
+              } else if (
+                  reduction == SegmentReductionType::MEAN ||
+                  reduction == SegmentReductionType::SUM) {
+                initial_value = 0;
+              } else if (reduction == SegmentReductionType::MIN) {
+                initial_value = std::numeric_limits<scalar_t>::infinity();
+              } else if (reduction == SegmentReductionType::PROD) {
+                initial_value = 1;
+              }
+
+              // ===== step2: apply reduction
+              for (const auto j : c10::irange(segment_length)) {
+                int64_t data_index = outer_idx * data.stride(axis) * data.size(axis)
+                                     + (lengths_cum_sum + j) * data.stride(axis) + inner_idx;
+                const auto val = values_data[data_index];
+                if (reduction == SegmentReductionType::MAX) {
+                  initial_value = at::_isnan(val)
+                      ? val
+                      : std::max<scalar_t>(initial_value, val);
+                } else if (
+                    reduction == SegmentReductionType::MEAN ||
+                    reduction == SegmentReductionType::SUM) {
+                  initial_value = initial_value + val;
+                } else if (reduction == SegmentReductionType::MIN) {
+                  initial_value = at::_isnan(val)
+                      ? val
+                      : std::min<scalar_t>(initial_value, val);
+                } else if (reduction == SegmentReductionType::PROD) {
+                  initial_value = initial_value * val;
+                }
+              }
+
+              // ===== step3: finalize reduction
+              TORCH_CHECK(segment_length >= 0);
+
+              if (segment_length == 0 && !initial.has_value() &&
```

Comment:
question:

Match pytorch_scatter and fill with 0s for empty segments or preserve current behavior?